### DISCUSSION TITLE
Naming convention for consistency วิธีการตั้งชื่อไฟล์

### DISCUSSION THREAD

**Discussion by bact** [PyThaiNLP/pythainlp#722] [General]
ปัจจุบันไฟล์ใน pythainlp/corpus มีรูปแบบการตั้งชื่อที่ไม่สม่ำเสมอ
เสนอให้มีการใช้ชื่อที่สม่ำเสมอครับ เพื่อความสะดวกในการดูแลโค้ด (มีความคาดเดาได้บางอย่าง) 

ตัวอย่าง

# การใช้ _ หรือ - คั่นคำ
- thaipos.py <- ไม่มีคั่น
- thaipos.json <- ไม่มีคั่น, ชื่อไฟล์ข้อมูลเหมือนไฟล์โค้ด
- thaiword.py <- ไม่มีคั่น
- thaiword.txt <- ไม่มีคั่น, ชื่อไฟล์ข้อมูลเหมือนไฟล์โค้ด
- thaisyllable.py <- ไม่มีคั่น
- thai_syllable.txt <- มี _ คั่น, ชื่อไฟล์ข้อมูลไม่เหมือนไฟล์โค้ด (ต่างตรงการคั่น)
- new-thaidict.txt <- มี - คั่น, ชื่อไฟล์ข้อมูลไม่เหมือนไฟล์โค้ด (ใช้คนละคำไปเลย dict vs word)
- newthaiword.py <- ไม่มีคั่น

# เอกพจน์หรือพหูพจน์
- country.py <-- เอกพจน์
- provinces.py <-- พหูพจน์
- stopwords.py <- พหูพจน์
- thaiword.py <- เอกพจน์

# suffix/prefix แสดงภาษา
- stopwords.py <- โค้ด ไม่ระบุภาษา
- stopwords-th.txt <- ชุดข้อมูล ระบุภาษาด้วย th
- thaiword.py <- โค้ด ระบุภาษาด้วย thai
- thaiword.txt <- ชุดข้อมูล ระบุภาษาด้วย thai

**Comment by cstorm125** [comment]
ผมคิดว่า

1. ใช้ _ เพราะมีปัญหาน้อยกว่า ไม่ถูกมองผิดเป็นเครื่องหมายลบ
2. ควรเป็นพหูจน์เพราะไฟล์นั้นๆมีสิ่งของมากกว่าหนึ่ง
3. suffix เพราะส่วนใหญ่เป็น thai เวลาเรากดหาน่าจะอยากหา functional keyword มากกว่า
3.1 เห็นว่าควรใช้ th เพราะตัวหนังสือน้อยกว่า

**Comment by bact** [comment]
ผมลองตามที่ @cstorm125 เสนอครับ ใช้ _ (underscore), ใช้พหูพจน์, ใช้ suffix _th ต่อท้ายถ้าเป็นข้อมูลภาษาไทย

**Comment by cstorm125** [comment]
Can we move these rules somewhere so we can close this?

**Comment by p16i** [comment]
maybe wiki?

**Comment by wannaphong** [comment]
Is it done? @bact

**Comment by kit1211** [comment]
Hi @wannaphong @bact @cstorm125

ตอบคำถาม Is it done? จากที่ตรวจโฟลเดอร์ pythainlp/corpus ใน main branch ปัจจุบัน

สรุป: ทำไปแล้ว **บางส่วน** แต่ยังไม่ครบ 100% และยังไม่ได้ย้ายกติกาไปไว้ในเอกสารอย่างที่ @cstorm125 กับ @p16i เสนอ

กติกาที่ทีมตกลงกันใน thread นี้ (2018):

1. ใช้ underscore `_` คั่นคำ ไม่ใช้ hyphen
2. ใช้พหูพจน์เมื่อไฟล์มีหลายรายการ เช่น stopwords, countries, provinces
3. ใช้ suffix `_th` ต่อท้ายไฟล์ข้อมูลภาษาไทย

ตัวอย่างที่ทำตามแล้วใน corpus ปัจจุบัน:

stopwords_th.txt
words_th.txt
countries_th.txt
thailand_provinces_th.csv
person_names_female_th.txt
syllables_th.txt
profanity_th.txt

ตัวอย่างที่ยังไม่สม่ำเสมอ (legacy):

tnc_freq.txt และ phupha_word_freqs.txt ไม่มี suffix _th
pos_orchid_perceptron.json ใช้ underscore แต่ไม่มี _th
blackboard-cls_v1.0.crfsuite ยังใช้ hyphen
ชื่อโมดูล thaiword รวมเข้า words_th แล้ว แต่บางไฟล์ model ยังไม่ตาม pattern เดียวกัน

ข้อเสนอเพื่อปิด discussion นี้:

1. เพิ่มหัวข้อ Corpus file naming ใน Wiki หรือ CONTRIBUTING.md อ้างอิงกติกา 3 ข้อด้านบน
2. ไฟล์ใหม่ทุกไฟล์ใน corpus ต้องตาม pattern `{purpose}_th.{ext}` หรือ `{purpose}_{lang}.{ext}`
3. ไฟล์เก่า rename ทีละ PR เมื่อมี breaking change window ไม่จำเป็นต้อง rename ทั้งหมดในครั้งเดียว
4. Mark discussion นี้เป็น Answered แล้วลิงก์ไปเอกสาร

CONTRIBUTING.md ปัจจุบันอ้างอิง namingconvention.org สำหรับ Python code แต่ยังไม่มี section เฉพาะ corpus data files ซึ่งเป็น gap ที่ discussion นี้พยายามแก้

ถ้า maintainer เห็นด้วย ผมแนะนำให้ @bact หรือ @wannaphong เปิด PR เพิ่ม section สั้นๆ ใน CONTRIBUTING.md แล้วปิด thread นี้ได้เลยครับ
